packages/sdk/src/db_connection_builder.ts

```diff
@@ -0,0 +1,260 @@
+    if (!/^wss?:/.test(this.#uri.protocol)) {
+      url.protocol = 'ws:';
+    }
+
+    const connection = new DBConnectionImpl(
+      this.spacetimeModule,
+      this.#emitter
+    );
+    connection.identity = this.#identity;
+    connection.token = this.#token;
+
+    let clientAddress = connection.clientAddress.toHexString();
+    url.searchParams.set('client_address', clientAddress);
+
+    connection.wsPromise = connection
+      .createWSFn({
+        url,
+        // TODO: This update this to bsatn
+        wsProtocol: 'v1.bin.spacetimedb',
+        authToken: connection.token,
+      })
+      .then(v => {
+        connection.ws = v;
+
+        connection.ws.onclose = () => {
+          this.#emitter.emit('disconnect', connection);
+        };
+        connection.ws.onerror = (e: ErrorEvent) => {
+          this.#emitter.emit('connectError', connection, e);
+        };
+        connection.ws.onopen = connection.handleOnOpen.bind(this);
+        connection.ws.onmessage = connection.handleOnMessage.bind(this);
+
+        return v;
+      })
+      .catch(e => {
+        stdbLogger('error', 'Error connecting to SpacetimeDB WS');
+        connection.on('connectError', e);
+        // TODO(cloutiertyler): I don't know but this makes it compile and
+        // I don't have time to investigate how to do this properly.
+        throw e;
+      });
+
+    return this.dbConnectionConstructor(connection);
+  }
+
+  /**
+   * Register a callback to be invoked upon authentication with the database.
+   *
+   * @param token The credentials to use to authenticate with SpacetimeDB.
+   * @param identity A unique identifier for a client connected to a database.
+   *
+   * The callback will be invoked with the `Identity` and private authentication `token` provided by the database to identify this connection.
+   *
+   * If credentials were supplied to connect, those passed to the callback will be equivalent to the ones used to connect.
+   *
+   * If the initial connection was anonymous, a new set of credentials will be generated by the database to identify this user.
+   *
+   * The credentials passed to the callback can be saved and used to authenticate the same user in future connections.
+   *
+   * @example
+   *
+   * ```ts
+   * spacetimeDBClient.onConnect((token, identity) => {
+   *  console.log("Connected to SpacetimeDB");
+   *  console.log("Token", token);
+   *  console.log("Identity", identity);
+   * });
+   * ```
+   */
+  onConnect(
+    callback: (
+      connection: DBConnectionImpl,
+      identity: Identity,
+      token: string
+    ) => void
+  ): DBConnectionBuilder<DBConnection> {
+    this.#emitter.on('connect', callback);
+    return this;
+  }
+
+  /**
+   * Register a callback to be invoked upon an error.
+   *
+   * @example
+   *
+   * ```ts
+   * spacetimeDBClient.onError((...args: any[]) => {
+   *  stdbLogger("warn","ERROR", args);
+   * });
+   * ```
+   */
+  onConnectError(
+    callback: (...args: any[]) => void
```

Comment:
Args should be `error: Error` or `error: any`.

Reply:
Ack, thank you

Reply:
Done! I've implemented this.